pyuvdata/uvfits.py

```diff
@@ -104,113 +87,272 @@ def read_uvfits(self, filename, run_check=True, check_extra=True,
+        Separated from full read so that header, metadata and data can be read independently.
+        """
+
+        if self.time_array is None:
+            # random parameter data has not been read in, do that first
+            self._get_parameter_data(vis_hdu)
+
+        # figure out what data to read in
+        blt_inds, freq_inds, pol_inds, history_update_string = \
+            self._select_preprocess(antenna_nums, antenna_names, ant_str, ant_pairs_nums,
+                                    frequencies, freq_chans, times, polarizations, blt_inds)
+
+        if blt_inds is not None:
+            blt_frac = len(blt_inds) / float(self.Nblts)
+        else:
+            blt_frac = 1
+
+        if freq_inds is not None:
+            freq_frac = len(freq_inds) / float(self.Nfreqs)
+        else:
+            freq_frac = 1
+
+        if pol_inds is not None:
+            pol_frac = len(pol_inds) / float(self.Npols)
+        else:
+            pol_frac = 1
+
+        min_frac = np.min([blt_frac, freq_frac, pol_frac])
+
+        if min_frac == 1:
+            # no select, read in all the data
+            if vis_hdu.header['NAXIS'] == 7:
+                raw_data_array = vis_hdu.data.data[:, 0, 0, :, :, :, :]
+                assert(self.Nspws == raw_data_array.shape[1])
+
+            else:
+                # in many uvfits files the spw axis is left out,
+                # here we put it back in so the dimensionality stays the same
+                raw_data_array = vis_hdu.data.data[:, 0, 0, :, :, :]
+                raw_data_array = raw_data_array[:, np.newaxis, :, :]
+        else:
+            # do select operations on everything except data_array, flag_array and nsample_array
+            self._select_metadata(blt_inds, freq_inds, pol_inds, history_update_string)
+
+            # just read in the right portions of the data and flag arrays
+            if blt_frac == min_frac:
+                if vis_hdu.header['NAXIS'] == 7:
+                    raw_data_array = vis_hdu.data.data[blt_inds, 0, 0, :, :, :, :]
+                    assert(self.Nspws == raw_data_array.shape[1])
+                else:
+                    # in many uvfits files the spw axis is left out,
+                    # here we put it back in so the dimensionality stays the same
+                    raw_data_array = vis_hdu.data.data[blt_inds, 0, 0, :, :, :]
+                    raw_data_array = raw_data_array[:, np.newaxis, :, :, :]
+                if freq_frac < 1:
+                    raw_data_array = raw_data_array[:, :, freq_inds, :, :]
+                if pol_frac < 1:
+                    raw_data_array = raw_data_array[:, :, :, pol_inds, :]
+            elif freq_frac == min_frac:
+                if vis_hdu.header['NAXIS'] == 7:
+                    raw_data_array = vis_hdu.data.data[:, 0, 0, :, freq_inds, :, :]
+                    # for some reason the axes get transposed. fix the ordering
```

Comment:
Why were the axes transposed? Is it because we're adding these axes that weren't on disk?

Reply:
I really don't understand why they were transposed. But it seemed to be consistently transposed, so I corrected for it.

Reply:
Okay. I'm a little uncomfortable not knowing the base reason for the transpose, but I suppose as long as the unit tests pass so we can catch any changes in behavior down the road, I'm okay with it.

Reply:
I'm somewhat uncomfortable about this too. I'm not sure how to figure it out. The one thing that made me feel better is that the check method that gets run by default after the read checks that the array shapes are correct. Of course that's not sufficient if different axes have the same length. Let me do a little more googling...